mailpoet/lib/Automation/Engine/Storage/AutomationRunStorage.php

```diff
@@ -41,14 +73,39 @@ public function getAutomationRun(int $id): ?AutomationRun {
    */
   public function getAutomationRunsForAutomation(Automation $automation): array {
     $table = esc_sql($this->table);
-    $query = (string)$this->wpdb->prepare("SELECT * FROM $table WHERE automation_id = %d", $automation->getId());
-    $result = $this->wpdb->get_results($query, ARRAY_A);
-    return is_array($result) ? array_map(
-      function(array $runData): AutomationRun {
+    $subjectTable = esc_sql($this->subjectTable);
+    $query = (string)$this->wpdb->prepare("SELECT * FROM $table WHERE automation_id = %d order by id", $automation->getId());
+    $automationRuns = $this->wpdb->get_results($query, ARRAY_A);
+    if (!is_array($automationRuns) || !$automationRuns) {
+      return [];
+    }
+
+    $automationRunIds = array_map(
+      function(array $runData): int {
+        return (int)$runData['id'];
+      },
+      $automationRuns
+    );
+
+    $sql = sprintf("SELECT * FROM $subjectTable WHERE automation_run_id in (%s) order by automation_run_id", implode(',', array_map(function() {return '%d';
+
+    }, $automationRunIds)));
+
+    $query = (string)$this->wpdb->prepare($sql, ...$automationRunIds);
+    $subjects = $this->wpdb->get_results($query, ARRAY_A);
+
+    return array_map(
+      function(array $runData) use ($subjects): AutomationRun {
+        $runData['subjects'] = array_filter(
```

Comment:
Be careful with `array_filter` as it makes gaps in the array indexes and thus a numeric array becomes associative. The safe path is to always wrap them with `array_values`. Or maybe we could simplify the search logic by preparing a map first...

Reply:
7188de08f